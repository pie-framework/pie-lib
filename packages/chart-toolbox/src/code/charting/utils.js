import { scaleBand, scalePoint } from '@vx/scale';
import { utils } from '../plot';

export const tickCount = utils.tickCount;
export const bounds = utils.bounds;
export const point = utils.point;

export const bandKey = (d, index) => `${index}-${d.label || '-'}`;

export const dataToXBand = (scaleX, data, width, type) => {
  switch (type) {
    case 'bar':
    case 'dotPlot':
    case 'linePlot':
      return scaleBand({
        rangeRound: [0, width],
        domain: data && data.map(bandKey),
        padding: 0.2,
      });

    case 'histogram':
      return scaleBand({
        rangeRound: [0, width],
        domain: data && data.map(bandKey),
        padding: 0,
      });

    case 'lineCross':
    case 'lineDot':
      return scalePoint({
        domain: data && data.map(bandKey),
        rangeRound: [0, width],
      });

    default:
      return scaleBand({
        range: [0, width],
        domain: data && data.map(bandKey),
        padding: 0,
      });
  }
};

export const getTickValues = (prop = {}) => {
  const tickValues = [];
  let tickVal = prop.min;

  while (tickVal <= prop.max) {
    tickValues.push(tickVal);
    tickVal = Math.round((tickVal + prop.step) * 10000) / 10000;
  }

  return tickValues;
};

export const getDomainAndRangeByChartType = (domain, range, chartType) => {
  let { step, labelStep, min, max } = range || {};

  if (!min) {
    min = 0;
  }

  if (!max || max < 0) {
    max = range.min + 1;
  }

  if (!step) {
    step = labelStep || 1;
  }
  if (!labelStep || (isNaN(labelStep) && step)) {
    labelStep = step || 1;
  }

  range.max = max;

  switch (chartType) {
    // if chart is dot plot or line plot, we should ignore step and make sure that min & max are integer values
    case 'dotPlot':
    case 'linePlot': {
      const intMin = Math.round(min);
      const intMax = Math.round(max);

      return {
        domain: {
          ...domain,
          step: 1,
          labelStep: 1,
          min: 0,
          max: 1,
        },
        range: {
          ...range,
          min: intMin,
          max: intMin === intMax ? intMin + 1 : intMax,
          labelStep,
          step: 1,
        },
      };
    }

    default:
      return {
        domain: {
          ...domain,
          step: 1,
          labelStep: 1,
          min: 0,
          max: 1,
        },
        range: {
          ...range,
          labelStep,
          step,
        },
      };
  }
};

export const getGridLinesAndAxisByChartType = (range, chartType) => {
  switch (chartType) {
    case 'lineDot':
    case 'lineCross':
      return {
        verticalLines: undefined,
        horizontalLines: getTickValues(range),
        leftAxis: true,
      };

    case 'dotPlot':
    case 'linePlot':
      return {
        verticalLines: [],
        horizontalLines: [],
        leftAxis: false,
      };

    default:
      return {
        verticalLines: [],
        horizontalLines: getTickValues(range),
        leftAxis: true,
      };
  }
};

export const getRotateAngle = (fontSize, height) => {
  if (height >= fontSize * 2) {
    return 25;
  }

  return 0;
};

export const getTopPadding = (barWidth) => {
  if (barWidth < 30) {
    return 50;
  }

  if (barWidth < 40) {
    return 30;
  }

  if (barWidth < 60) {
    return 15;
  }

  return 0;
};

// This function calculates the transformation scale for SVG and the icon's vertical distance from its category
export const getScale = (width) => {
  let scale, deltay;

  if (width > 91) {
    scale = 1.3;
    deltay = -55;
  } else if (width > 45) {
    scale = 1.1;
    deltay = -45;
  } else if (width > 40) {
    scale = 0.5 + (width - 34) * 0.02;
    deltay = -25;
  } else if (width > 30) {
    scale = 0.5 + (width - 34) * 0.02;
    deltay = -20;
  } else {
    scale = 0.5 * Math.pow(0.98, 34 - width); // 0.98 is the reduction factor. Adjust to control scaling.
    deltay = -15;
  }

  return { scale, deltay };
};

export const getAdjustedX = (width, scaleValue) => {
  const innerWidthOriginal = 57;
  const effectiveInnerWidth = innerWidthOriginal * scaleValue;
  return (width - effectiveInnerWidth) / 2;
};
