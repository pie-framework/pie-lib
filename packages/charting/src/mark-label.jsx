import React, { useState, useCallback, useEffect, useRef } from 'react';
import cn from 'classnames';
import { withStyles } from '@material-ui/core/styles';
import AutosizeInput from 'react-input-autosize';
import PropTypes from 'prop-types';
import { types } from '@pie-lib/plot';
import { correct, incorrect, disabled } from './common/styles';
import { color } from '@pie-lib/render-ui';
import { renderMath } from '@pie-lib/math-rendering';

const styles = (theme) => ({
  input: {
    float: 'right',
    fontFamily: theme.typography.fontFamily,
    fontSize: theme.typography.fontSize,
    border: 'none',
    color: color.primaryDark(),
    '&.error': { border: `2px solid ${theme.palette.error.main}` },
    '&.correct': correct('color'),
    '&.incorrect': incorrect('color'),
    '&.disabled': {
      ...disabled('color'),
      backgroundColor: 'transparent !important',
    },
  },
});

function isFractionFormat(label) {
  const trimmedLabel = label?.trimStart().trimEnd() || '';
  const stringRegex = /[a-zA-Z]/;
  // Check if label contain at least one letter
  if (stringRegex.test(label)) {
    return false;
  }
  const fracRegex = new RegExp(/[1-9][0-9]*\/[1-9][0-9]*$/g);
  return !!trimmedLabel?.match(fracRegex) || false;
}

function getLabelMathFormat(label) {
  const trimmedLabel = label?.trimStart().trimEnd() || '';
  let fraction;
  let mixedNr = '';
  let improperFraction = trimmedLabel.split(' ');
  if (improperFraction[1] && improperFraction[1].includes('/')) {
    fraction = improperFraction[1].split('/') || '';
  } else {
    fraction = trimmedLabel?.split('/') || '';
  }

  let formattedLLabel;
  if (isFractionFormat(label)) {
    if (improperFraction[0] && improperFraction[1]) {
      mixedNr = improperFraction[0];
    }
    formattedLLabel = `\\(${mixedNr}\\frac{${fraction[0]}}{${fraction[1]}}\\)`;
    return formattedLLabel;
  }
  return undefined;
}

export const MarkLabel = (props) => {
  // eslint-disable-next-line no-unused-vars
  const [input, setInput] = useState(null);
  const _ref = useCallback((node) => setInput(node), null);

  const {
    mark,
    classes,
    disabled,
    inputRef: externalInputRef,
    barWidth,
    rotate,
    correctness,
    autoFocus,
    error,
  } = props;

  const [label, setLabel] = useState(mark.label);
  const [mathLabel, setMathLabel] = useState(getLabelMathFormat(mark.label));
  const [isEditing, setIsEditing] = useState(false);
  // let root = useRef(null);
  const onChange = (e) => {
    setLabel(e.target.value);
  };

  const isMathRendering = () => {
    return isEditing === false && mathLabel !== undefined;
  };

  const onChangeProp = (e) => {
    setMathLabel(getLabelMathFormat(mark.label));
    setIsEditing(false);
    props.onChange(e.target.value);
  };
  let extraStyle = {};

  if (rotate) {
    extraStyle = {
      width: 'unset',
      textAlign: 'left',
    };
  }

  // useState only sets the value once, to synch props to state need useEffect
  useEffect(() => {
    setLabel(mark.label);
  }, [mark.label]);

  return isMathRendering() ? (
    <div
      dangerouslySetInnerHTML={{ __html: getLabelMathFormat(label) }}
      onClick={() => setIsEditing(true)}
      style={{
        pointerEvents: 'auto',
        textAlign: 'center',
        minWidth: barWidth,
        fontSize: '14px',
        fontFamily: 'Roboto',
        color: '#283593',
      }}
    ></div>
  ) : (
    <AutosizeInput
      inputRef={(r) => {
        _ref(r);
        externalInputRef(r);
      }}
      autoFocus={isEditing || autoFocus}
      disabled={disabled}
      inputClassName={cn(classes.input, correctness && correctness.label, disabled && 'disabled', error && 'error')}
      inputStyle={{
        minWidth: barWidth,
        textAlign: 'center',
        background: 'transparent',
        boxSizing: 'border-box',
        paddingLeft: 0,
        paddingRight: 0,
        ...extraStyle,
      }}
      value={label}
      style={{
        position: 'fixed',
        pointerEvents: 'auto',
        top: 0,
        left: 0,
        minWidth: barWidth,
        transformOrigin: 'left',
        transform: `rotate(${rotate}deg)`,
      }}
      onChange={onChange}
      onBlur={onChangeProp}
    />
  );
};

MarkLabel.propTypes = {
  disabled: PropTypes.bool,
  onChange: PropTypes.func,
  graphProps: types.GraphPropsType,
  classes: PropTypes.object,
  inputRef: PropTypes.func,
  mark: PropTypes.object,
  barWidth: PropTypes.number,
  rotate: PropTypes.number,
  correctness: PropTypes.shape({
    value: PropTypes.string,
    label: PropTypes.string,
  }),
};

export default withStyles(styles)(MarkLabel);
