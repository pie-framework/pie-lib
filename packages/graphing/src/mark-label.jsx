import React, { useState, useCallback, useEffect } from 'react';
import cn from 'classnames';
import { withStyles } from '@material-ui/core/styles';
import AutosizeInput from 'react-input-autosize';
import PropTypes from 'prop-types';
import { GraphPropsType } from '@pie-lib/plot/lib/types';
import { useDebounce } from './use-debounce';

const styles = theme => ({
  input: {
    float: 'right',
    padding: theme.spacing.unit * 0.5,
    fontFamily: theme.typography.fontFamily,
    fontSize: theme.typography.fontSize,
    border: `solid 1px ${theme.palette.secondary.main}`,
    borderRadius: '3px',
    color: theme.palette.primary.dark
  },
  disabled: {
    border: `solid 1px ${theme.palette.primary.dark}`,
    background: 'white'
  }
});

export const position = (graphProps, mark, rect) => {
  rect = rect || { width: 0, height: 0 };
  const { scale, domain, range } = graphProps;

  const rightEdge = scale.x(mark.x) + rect.width;
  const bottomEdge = scale.y(mark.y) + rect.height;

  const h = rightEdge >= scale.x(domain.max) ? 'left' : 'right';
  const v = bottomEdge >= scale.y(range.min) ? 'top' : 'bottom';
  return `${v}-${h}`;
};

export const coordinates = (graphProps, mark, rect, position) => {
  const { scale } = graphProps;
  const shift = 10;
  rect = rect || { width: 0, height: 0 };
  switch (position) {
    case 'bottom-right': {
      return { left: scale.x(mark.x) + shift, top: scale.y(mark.y) + shift };
    }
    case 'bottom-left': {
      return { left: scale.x(mark.x) - shift - rect.width, top: scale.y(mark.y) + shift };
    }
    case 'top-left': {
      return {
        left: scale.x(mark.x) - shift - rect.width,
        top: scale.y(mark.y) - shift - rect.height
      };
    }
    case 'top-right': {
      return {
        left: scale.x(mark.x) + shift,
        top: scale.y(mark.y) - shift - rect.height
      };
    }
  }
};

export const MarkLabel = props => {
  const [input, setInput] = useState(null);
  const _ref = useCallback(node => setInput(node));

  const { mark, graphProps, classes, disabled, inputRef: externalInputRef } = props;
  const [label, setLabel] = useState(mark.label);

  const onChange = e => setLabel(e.target.value);

  const debouncedLabel = useDebounce(label, 200);

  // useState only sets the value once, to synch props to state need useEffect
  useEffect(() => {
    setLabel(mark.label);
  }, [mark.label]);

  // pick up the change to debouncedLabel and save it
  useEffect(() => {
    if (typeof debouncedLabel === 'string' && debouncedLabel !== mark.label) {
      props.onChange(debouncedLabel);
    }
  }, [debouncedLabel]);

  const rect = input ? input.getBoundingClientRect() : { width: 0, height: 0 };
  const pos = position(graphProps, mark, rect);
  const leftTop = coordinates(graphProps, mark, rect, pos);

  const style = {
    position: 'absolute',
    pointerEvents: 'auto',
    ...leftTop
  };

  return (
    <AutosizeInput
      inputRef={r => {
        _ref(r);
        externalInputRef(r);
      }}
      disabled={disabled}
      inputClassName={cn(classes.input, disabled && classes.disabled)}
      value={label}
      style={style}
      onChange={onChange}
    />
  );
};

MarkLabel.propTypes = {
  disabled: PropTypes.bool,
  onChange: PropTypes.func,
  graphProps: GraphPropsType,
  classes: PropTypes.object,
  inputRef: PropTypes.func,
  mark: PropTypes.object
};

export default withStyles(styles)(MarkLabel);
