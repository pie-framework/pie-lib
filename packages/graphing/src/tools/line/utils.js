import debug from 'debug';
// import { swap } from '../point/utils';

const log = debug('pie-lib:charting:line:utils');

const lineExpressionRegex = new RegExp(/([+|-]?\d+)?(x)?([+|-]\d+)?/i);

export class Expression {
  constructor(multiplier, b) {
    this.multiplier = multiplier;
    this.isVerticalLine = Math.abs(this.multiplier) === Infinity;
    this.b = this.isVerticalLine ? undefined : b;
  }

  expr() {
    return this.isVerticalLine ? 'x' : `${this.multiplier}x + ${this.b}`;
  }

  getY(x) {
    return this.isVerticalLine ? x : this.multiplier * x + this.b;
  }

  getX(y) {
    return this.isVerticalLine ? y : (y - this.b) / this.multiplier;
  }

  equals(other) {
    return this.multiplier === other.multiplier && this.b === other.b;
  }
}

export const pointsHaveSameExpression = (a, b) => {
  const expressionA = expression(a.from, a.to);
  const expressionB = expression(b.from, b.to);
  return expressionA.equals(expressionB);
};
/**
 * Create a linear expression from 2 points
 */

export const expression = (from, to) => {
  log('[expression] from: ', from, 'to: ', to);
  const multiplier = (to.y - from.y) / (to.x - from.x);
  const zeroedY = multiplier * from.x;
  const b = from.y - zeroedY;
  return new Expression(multiplier, b);
};

export const point = (x, y) => ({ x, y });

export const pointsFromExpression = expression => {
  const to = point(0, expression.b);
  const huh = expression.getX(0);
  const from = point(huh, 0);

  return {
    from,
    to
  };
};

export const expressionFromDescriptor = descriptor => {
  const lineDescriptor = descriptor.match(lineExpressionRegex);
  const maybeB = parseInt(lineDescriptor[3], 10);
  const maybeMultiplier = parseInt(lineDescriptor[1], 10);
  let multiplier = isNaN(maybeMultiplier)
    ? lineDescriptor[2] === undefined
      ? 0
      : 1
    : maybeMultiplier;
  let b = isNaN(maybeB) ? 0 : maybeB;

  // it's a constant, no variable found
  if (lineDescriptor[2] === undefined) {
    b = multiplier;
    multiplier = 0;
  }

  return new Expression(multiplier, b);
};

export const hasLine = (lines, line) => lineIndex(lines, line) !== -1;

export const removeLine = (lines, line) => {
  const index = lineIndex(lines, line);
  if (index === -1) {
    return lines;
  }
  const out = lines.slice();
  out.splice(index, 1);
  return out;
};

export const removeLines = (lines, toRemove) => {
  log('[removeLines] points: ', lines, toRemove);
  toRemove = Array.isArray(toRemove) ? toRemove : [toRemove];
  const out = lines.filter(p => !hasLine(toRemove, p));
  log('[removePoints] points: ', out);
  return out;
};

export const pointsEqual = (a, b) => a.x == b.x && a.y == b.y;

export const linesEqual = (a, b) => pointsEqual(a.from, b.from) && pointsEqual(a.to, b.to);

export const lineIndex = (lines, line) => lines.findIndex(l => linesEqual(l, line));

// export const swapLine = (lines, from, to) => swap(lines, from, to, lineIndex);
