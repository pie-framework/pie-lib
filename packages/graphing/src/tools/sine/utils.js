import _ from 'lodash';
import debug from 'debug';
import Point from '@mapbox/point-geometry';
const log = debug('pie-lib:graphing;sine:utils');

/**
 * build the x points with root as the anchor.
 * adds padding left and right so that
 * the anchor can be moved left/right
 * and there will be points off stage rendering.
 *
 *  root: 0, freq: 2, min: -5, max: 5 ==> -5,-4-2,0,2,4,5
 */
export const xPoints = (root, freq, min, max) => {
  return _.concat(
    min - freq,
    _.rangeRight(root, min - freq, freq * -1),
    _.range(root + freq, max + freq, freq),
    max + freq
  );
};

export const sinY = (amplitude, freq) => x => {
  const num = 2 * Math.PI * x;
  const frac = num / freq;
  return amplitude * parseFloat(Math.sin(frac).toFixed(10));
};

export const buildDataPoints = (min, max, root, interval, yFn) => {
  log('[buildDataPoints] min:', min, 'max:', max, 'root:', root);
  // const fn = sinY(amplitude, freq);
  const diff = max - root.x;
  const plusDiff = min - root.x;
  const xs = xPoints(0, interval, min - root.x - diff, max - root.x - plusDiff);
  log('xs:', xs);
  return xs.map(v => new Point(v, yFn(v))).map(p => p.add(new Point(root.x, root.y)));
};
