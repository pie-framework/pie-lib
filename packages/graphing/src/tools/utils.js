import debug from 'debug';
import Point from '@mapbox/point-geometry';
import _ from 'lodash';
const log = debug('pie-lib:graphing:tools:utils');
/**
 * build the x points with root as the anchor.
 */
export const xPoints = (root, freq, min, max) => {
  freq = Math.abs(freq);
  log('[xPoints] root:', root, 'freq:', freq, ' min:', min, ' max:', max);
  const leftpoints = _.rangeRight(root, min - freq, freq * -1);
  const rightpoints = _.range(root + freq, max + freq, freq);

  return _.concat(leftpoints, rightpoints);
};

/**
 * Calculate y for x value using amp, freq and shift
 * https://www.desmos.com/calculator/f1psfoiiv6
 */
export const sinY = (amplitude, freq, shift) => x => {
  shift = { phase: 0, vertical: 0, ...shift };
  const TWO_PI = Math.PI * 2;
  const num = TWO_PI * (x - shift.phase);
  const frac = num / freq;
  const sin = Math.sin(frac);
  const sinFloat = parseFloat(sin).toFixed(10);
  return amplitude * sinFloat + shift.vertical;
};

/**
 * You need three points for the quadratic eq, so we mirror the edge point
 * https://www.desmos.com/calculator/xhz1xqgpca
 */
export const parabolaFromTwoPoints = (root, edge) => {
  edge = edge || { ...root };
  const mirrorEdge = { x: root.x - (edge.x - root.x), y: edge.y };
  return parabolaFromThreePoints(root, edge, mirrorEdge);
};

export const parabolaFromThreePoints = (one, two, three) => {
  log(one, two, three);
  const { a, b, c } = pointsToABC(one, two, three);
  log(a, b, c);
  return parabola(a, b, c);
};
/** derive a,b,c from the 3 points */
export const pointsToABC = (one, two, three) => {
  const a1 = Math.pow(one.x, 2) * -1 + Math.pow(two.x, 2);
  const b1 = one.x * -1 + two.x;
  const d1 = one.y * -1 + two.y;
  const a2 = Math.pow(two.x, 2) * -1 + Math.pow(three.x, 2);
  const b2 = two.x * -1 + three.x;
  const d2 = two.y * -1 + three.y;
  const bMult = (b2 / b1) * -1;
  const a3 = bMult * a1 + a2;
  const d3 = bMult * d1 + d2;
  const a = d3 / a3;
  const b = (d1 - a1 * a) / b1;
  const c = one.y - a * Math.pow(one.x, 2) - b * one.x;
  return { a, b, c };
};

/**
 * y=ax^2+bx+c
 */
export const parabola = (a, b, c) => x => a * Math.pow(x, 2) + b * x + c;

/**
 * Generate a set of data points, add spacing before min and after max if there is space between minx + min and maxX and max
 * @param {*} min
 * @param {*} max
 * @param {*} root
 * @param {*} edge
 * @param {*} interval
 * @param {*} yFn
 */
export const buildDataPoints = (min, max, root, edge, interval, yFn) => {
  log('[buildDataPoints] min:', min, 'max:', max, 'root:', root);
  edge = edge ? edge : { ...root };
  const minX = Math.min(root.x, edge.x);
  const maxX = Math.max(root.x, edge.x);
  const leftSpace = min - minX;
  const rightSpace = max - maxX;
  console.log(minX, interval);
  const xs = xPoints(minX, interval, min - rightSpace, max - leftSpace);
  log('[buildDataPoints]:xs:', xs);
  return xs.map(v => new Point(v, yFn(v)));
};
