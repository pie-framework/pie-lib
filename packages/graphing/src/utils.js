import head from 'lodash/head';
import tail from 'lodash/tail';
import { utils } from '@pie-lib/plot';
import invariant from 'invariant';
import isEqual from 'lodash/isEqual';

export const tickCount = utils.tickCount;
export const bounds = utils.bounds;
export const point = utils.point;

export const polygonToArea = points => {
  const h = head(points);
  const area = {
    left: h.x,
    top: h.y,
    bottom: h.y,
    right: h.x
  };
  return tail(points).reduce((a, p) => {
    a.left = Math.min(a.left, p.x);
    a.top = Math.max(a.top, p.y);
    a.bottom = Math.min(a.bottom, p.y);
    a.right = Math.max(a.right, p.x);
    return a;
  }, area);
};

export const lineToArea = (from, to) => pointsToArea(from, to);

export const pointsToArea = (a, b) => {
  invariant(!!a && !!b, 'a or b is undefined');
  const left = Math.min(a.x, b.x);
  const top = Math.max(a.y, b.y);
  const bottom = Math.min(a.y, b.y);
  const right = Math.max(a.x, b.x);
  return { left, top, bottom, right };
};

export const isDomainRangeEqual = (graphProps, nextGraphProps) => {
  return (
    isEqual(graphProps.domain, nextGraphProps.domain) &&
    isEqual(graphProps.range, nextGraphProps.range)
  );
};
