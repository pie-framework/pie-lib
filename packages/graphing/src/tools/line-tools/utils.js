import { calculateThirdPointOnLine } from '../../utils';

export const arrowDimensions = {
  none: 0,
  vector: 7,
  ray: 5,
  line: 5
};

export const calculateCorrectScaledPoints = (scale, from, to, type = 'none') => {
  const scaledFromX = scale.x(from.x);
  const scaledFromY = scale.y(from.y);
  let scaledToX = scale.x(to.x);
  let scaledToY = scale.y(to.y);
  let middlePoint;

  middlePoint = calculateThirdPointOnLine(
    { x: scaledToX, y: scaledToY },
    { x: scaledFromX, y: scaledFromY },
    {
      domain: {
        min: scale.x(to.x) - arrowDimensions[type],
        max: scale.x(to.x) + arrowDimensions[type]
      },
      range: {
        min: scale.y(to.y) - arrowDimensions[type],
        max: scale.y(to.y) + arrowDimensions[type]
      }
    }
  );

  if (middlePoint && isFinite(middlePoint.x) && isFinite(middlePoint.y)) {
    scaledToX = middlePoint.x;
    scaledToY = middlePoint.y;
  }

  return {
    x: scaledToX,
    y: scaledToY
  };
};
