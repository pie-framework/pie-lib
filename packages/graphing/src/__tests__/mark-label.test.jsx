import { shallow } from 'enzyme';
import React from 'react';
import { MarkLabel, position, coordinates } from '../mark-label';
import { graphProps as getGraphProps } from '../__tests__/utils';

const xyFn = () => {
  const out = jest.fn(n => n);
  out.invert = jest.fn(n => n);
  return out;
};

describe('MarkLabel', () => {
  let w;
  let onChange = jest.fn();
  const wrapper = extras => {
    const defaults = {
      classes: {},
      className: 'className',
      onChange,
      mark: { x: 1, y: 1 },
      graphProps: getGraphProps(0, 10, 0, 10)
    };
    const props = { ...defaults, ...extras };
    return shallow(<MarkLabel {...props} />);
  };
  describe('snapshot', () => {
    it('renders', () => {
      w = wrapper();
      expect(w).toMatchSnapshot();
    });
    it('renders', () => {
      w = wrapper({ mark: { x: 10, y: 10 } });
      expect(w).toMatchSnapshot();
    });
  });
});

describe('position', () => {
  const assertPosition = (mark, rect, expected) => {
    it(`${mark.x},${mark.y} + ${rect.width},${rect.height} => ${expected}`, () => {
      // we set range.min to a value because in pixels - the greater the Y the lower down on the screen.
      const graphProps = getGraphProps(0, 10, 10, 0);
      const result = position(graphProps, mark, rect);
      expect(result).toEqual(expected);
    });
  };

  assertPosition({ x: 0, y: 0 }, { width: 10, height: 10 }, 'top-left');
  assertPosition({ x: 0, y: 0 }, { width: 1, height: 1 }, 'bottom-right');
  assertPosition({ x: 0, y: 0 }, { width: 10, height: 0 }, 'bottom-left');
  assertPosition({ x: 0, y: 0 }, { width: 0, height: 10 }, 'top-right');
});

describe('coordinates', () => {
  const assertCoordinates = (mark, rect, pos, expected) => {
    it(`${mark.x}, ${mark.y} -> ${pos} = ${expected.left}, ${expected.top}`, () => {
      const result = coordinates(getGraphProps(), mark, rect, pos);
      expect(result).toEqual(expected);
    });
  };
  assertCoordinates({ x: 0, y: 0 }, { width: 0, height: 0 }, 'top-left', { left: -10, top: -10 });
  assertCoordinates({ x: 0, y: 0 }, { width: 0, height: 0 }, 'bottom-left', { left: -10, top: 10 });
  assertCoordinates({ x: 0, y: 0 }, { width: 0, height: 0 }, 'top-right', { left: 10, top: -10 });
  assertCoordinates({ x: 0, y: 0 }, { width: 0, height: 0 }, 'bottom-right', {
    left: 10,
    top: 10
  });
});
