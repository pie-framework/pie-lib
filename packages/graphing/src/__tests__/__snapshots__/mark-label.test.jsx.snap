// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarkLabel snapshot renders 1`] = `
<AutosizeInput
  injectStyles={true}
  inputClassName=""
  inputRef={[Function]}
  minWidth={1}
  onChange={[Function]}
  style={
    Object {
      "left": 11,
      "pointerEvents": "auto",
      "position": "absolute",
      "top": -9,
    }
  }
/>
`;

exports[`MarkLabel snapshot renders 2`] = `
<AutosizeInput
  injectStyles={true}
  inputClassName=""
  inputRef={[Function]}
  minWidth={1}
  onChange={[Function]}
  style={
    Object {
      "left": 0,
      "pointerEvents": "auto",
      "position": "absolute",
      "top": 0,
    }
  }
/>
`;
